# Autograd 自动求导
requires_grad为Ture时，Tensor会追踪记录对该Tensor的所有操作（.grad_fn属性会记录所有操作），当计算完成后，可以调用 .backward()来自动计算所有梯度，并自动累加到 .grad 属性。


In [3]:
import torch
x = torch.ones(2,2,requires_grad=True)
print(x)
# 做一次运算
y = x + 2
print(y)
print(y.grad_fn)
print(x.grad_fn)

tensor([[1., 1.],
        [1., 1.]], requires_grad=True)
tensor([[3., 3.],
        [3., 3.]], grad_fn=<AddBackward0>)
None


In [4]:
z = y * y * 3
out = z.mean()
print(z,out)

tensor([[27., 27.],
        [27., 27.]], grad_fn=<MulBackward0>) tensor(27., grad_fn=<MeanBackward0>)


In [5]:
out.backward()
print(x.grad)

tensor([[4.5000, 4.5000],
        [4.5000, 4.5000]])


标量是只包含一个元素的数据。

mean() 函数是求平均值所以out对于$x_i$的表达式如下：

$out = \frac{1}{4}\sum_{i=1}^4 z_i$，$z_i = 3(x_i+2)^2$ 

所以 $\frac{\sigma{out}}{\sigma{x}_i}=\frac{3}{2}(x_i+2)$，所以$x_i$取1时，$\frac{\sigma{out}}{\sigma{x}_i}= 4.5$